### PR TITLE
Fix compilation with clang-13.

### LINKED ISSUE #14004: compilation failure with clang 13 and grid generator
I am getting an internal clang error when compiling both master and the release candidate:

```
[drwells@archway build-clang]$ bash clang-error.sh
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /usr/bin/clang++ -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-psabi -Wfloat-conversion -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -Wno-pass-failed -pthread -Wno-unused-local-typedefs -Og -ggdb -Wa,--compress-debug-sections -DBOOST_NO_AUTO_PTR -DDEBUG -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/source/grid -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/include -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/include -isystem /home/drwells/Applications/petsc-dev-opt/include -isystem /usr/include/hypre -isystem /home/drwells/Applications/arborx/include -isystem /home/drwells/Applications/arborx/include/details -isystem /opt/p4est-deal-ii/FAST/include -fcolor-diagnostics -o gg.o -c /home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/source/grid/grid_generator.cc
1.      /home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/source/grid/grid_generator.cc:6148:3: current parser token 'template'
2.      /home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/source/grid/grid_generator.cc:39:1 <Spelling=/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/include/deal.II/base/config.h:73:32>: parsing namespace 'dealii'
3.      /home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/source/grid/grid_generator.cc:45:1: parsing namespace 'dealii::malloc_consolidate(): invalid chunk size
clang-error.sh: line 1: 16457 Aborted                 (core dumped) /usr/bin/clang++ -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-psabi -Wfloat-conversion -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -Wno-pass-failed -pthread -Wno-unused-local-typedefs -Og -ggdb -Wa,--compress-debug-sections -DBOOST_NO_AUTO_PTR -DDEBUG -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/source/grid -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/include -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/include -isystem /home/drwells/Applications/petsc-dev-opt/include -isystem /usr/include/hypre -isystem /home/drwells/Applications/arborx/include -isystem /home/drwells/Applications/arborx/include/details -isystem /opt/p4est-deal-ii/FAST/include -fcolor-diagnostics -o gg.o -c /home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/source/grid/grid_generator.cc
[drwells@archway build-clang]$ bash clang-error.sh
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /usr/bin/clang++ -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-psabi -Wfloat-conversion -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -Wno-pass-failed -pthread -Wno-unused-local-typedefs -Og -ggdb -Wa,--compress-debug-sections -DBOOST_NO_AUTO_PTR -DDEBUG -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/source/grid -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/include -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/include -isystem /home/drwells/Applications/petsc-dev-opt/include -isystem /usr/include/hypre -isystem /home/drwells/Applications/arborx/include -isystem /home/drwells/Applications/arborx/include/details -isystem /opt/p4est-deal-ii/FAST/include -fcolor-diagnostics -o gg.o -c /home/drwells/Documents/Code/CPP/dealii-dev/source/grid/grid_generator.cc
1.      /home/drwells/Documents/Code/CPP/dealii-dev/source/grid/grid_generator.cc:6148:3: current parser token 'template'
2.      /home/drwells/Documents/Code/CPP/dealii-dev/source/grid/grid_generator.cc:39:1 <Spelling=/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/include/deal.II/base/config.h:73:32>: parsing namespace 'dealii'
3.      /home/drwells/Documents/Code/CPP/dealii-dev/source/grid/grid_generator.cc:45:1: parsing namespace 'dealii::malloc(): unaligned tcache chunk detected
clang-error.sh: line 1: 16507 Aborted                 (core dumped) /usr/bin/clang++ -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-psabi -Wfloat-conversion -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -Wno-pass-failed -pthread -Wno-unused-local-typedefs -Og -ggdb -Wa,--compress-debug-sections -DBOOST_NO_AUTO_PTR -DDEBUG -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/source/grid -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/build-clang/include -I/home/drwells/Documents/Code/CPP/dealii-9.4.0-rc1/include -isystem /home/drwells/Applications/petsc-dev-opt/include -isystem /usr/include/hypre -isystem /home/drwells/Applications/arborx/include -isystem /home/drwells/Applications/arborx/include/details -isystem /opt/p4est-deal-ii/FAST/include -fcolor-diagnostics -o gg.o -c /home/drwells/Documents/Code/CPP/dealii-dev/source/grid/grid_generator.cc
```
This is with clang-13 on Arch Linux. I'm investigating.

### PR DESCRIPTION
**Pull request by drwells**
Fixes #14004. Shuffling things around and getting rid of the recursion seems to be enough to make clang happy.

### CODE REVIEW

**Comment by drwells** [comment]
Once this is approved we should cherry-pick onto 9.4.

**Review by kronbichler — APPROVED** [review]
This is indeed much nicer to read (I am the culprit for this code).

**Comment by kronbichler** [comment]
Hm, the tester is not happy:
```
The following tests FAILED:
	261 - particles/generators_03.debug (Failed)
	1386 - data_out/data_out_base_vtu_05.debug (Failed)
	1444 - data_out/data_out_base_vtk_05.debug (Failed)
	1448 - grid/concentric_hyper_shell_01.debug (Failed)
	1462 - grid/grid_hyper_shell_08.debug (Failed)
	1490 - grid/grid_generator_from_name_and_argument_01.debug (Failed)
	1494 - grid/grid_hyper_shell_05.debug (Failed)
	1498 - grid/grid_out_gnuplot_01.debug (Failed)
	1531 - grid/grid_hyper_shell_07.debug (Failed)
	1582 - grid/grid_out_gnuplot_02.debug (Failed)
	1585 - grid/find_all_active_cells_around_point_02.debug (Failed)
	1596 - grid/grid_hyper_shell_02.debug (Failed)
	1598 - grid/project_to_object_01.debug (Failed)
	1618 - grid/grid_generator_eccentric_hyper_shell.debug (Failed)
	1659 - grid/concentric_hyper_shell_02.debug (Failed)
	1666 - grid/grid_hyper_shell_03.debug (Failed)
	1749 - grid/grid_hyper_shell.debug (Failed)
	1769 - grid/grid_hyper_shell_04.debug (Failed)
	3376 - mappings/mapping_q_inverse_quadratic_approximation.debug (Failed)
	3379 - mappings/mapping_q_real_to_unit_internal.debug (Failed)
	3384 - mappings/mapping_fe_field_08.debug (Failed)
	3539 - manifold/spherical_manifold_10.debug (Failed)
	3543 - manifold/polar_manifold_01.debug (Failed)
	3564 - manifold/spherical_manifold_09.debug (Failed)
	3567 - manifold/transfinite_manifold_06.debug (Failed)
	3590 - manifold/tria_accessor_point_02.debug (Failed)
	3610 - manifold/tria_accessor_point_01.debug (Failed)
	3814 - matrix_free/point_evaluation_05.debug (Failed)
	3838 - matrix_free/point_evaluation_08.debug (Failed)
	3853 - matrix_free/point_evaluation_10.debug (Failed)
	3866 - matrix_free/point_evaluation_01.debug (Failed)
	3872 - matrix_free/point_evaluation_09.debug (Failed)
	3878 - matrix_free/point_evaluation_12.debug (Failed)
	3899 - matrix_free/point_evaluation_16.debug (Failed)
	3902 - matrix_free/point_evaluation_02.debug (Failed)
	3906 - matrix_free/point_evaluation_11.debug (Failed)
	3925 - matrix_free/point_evaluation_03.debug (Failed)
	3929 - matrix_free/point_evaluation_14.debug (Failed)
	3941 - matrix_free/point_evaluation_06.debug (Failed)
	3985 - matrix_free/point_evaluation_07.debug (Failed)
	3987 - matrix_free/point_evaluation_04.debug (Failed)
	4400 - numerics/no_flux_15.debug (Failed)
```

**Comment by drwells** [comment]
I must have made some silly mistake - I'll patch it.

File: source/grid/grid_generator.cc [hunk 1]
```diff
@@ -5927,6 +5927,206 @@ namespace GridGenerator
...
+      hyper_shell_12(Triangulation<3> &tria,
+                     const Point<3> &  p,
+                     const double      inner_radius,
+                     const double      outer_radius)
+      {
+        std::vector<Point<3>>    vertices;
+        std::vector<CellData<3>> cells;
+
+        const double irad = inner_radius / std::sqrt(3.0);
+        const double orad = outer_radius / std::sqrt(3.0);
```
Comment (kronbichler) [inline source/grid/grid_generator.cc:5996]: Here you divide again by `sqrt(3)`, which you have already done on the arguments passed in to this function.

Comment (drwells) [reply source/grid/grid_generator.cc:5998]: I fixed this and a problem with manifolds.

**Review by kronbichler — APPROVED** [review]

**Review by bangerth — APPROVED** [review]

**Pull request merged**